Source/WebKit/UIProcess/ios/forms/WKFileUploadPanel.mm

```diff
@@ -974,6 +1022,89 @@ - (void)_processMediaInfoDictionaries:(NSArray *)infos successBlock:(void (^)(NS
+}
+
+- (void)_processPickerResults:(NSArray<PHPickerResult *> *)results atIndex:(NSUInteger)index processedResults:(NSMutableArray<_WKFileUploadItem *> *)processedResults successBlock:(void (^)(NSArray<_WKFileUploadItem *> *processedResults))successBlock failureBlock:(void (^)(void))failureBlock
+{
+    NSUInteger count = [results count];
+    if (index == count) {
+        successBlock(processedResults);
+        return;
+    }
+
+    PHPickerResult *result = [results objectAtIndex:index];
+    ASSERT(index < count);
+    index++;
+
+    auto uploadItemSuccessBlock = ^(_WKFileUploadItem *uploadItem) {
+        [processedResults addObject:uploadItem];
+        [self _processPickerResults:results atIndex:index processedResults:processedResults successBlock:successBlock failureBlock:failureBlock];
+    };
+
+    [self _uploadItemFromResult:result successBlock:uploadItemSuccessBlock failureBlock:failureBlock];
+}
+
+- (void)_uploadItemFromResult:(PHPickerResult *)result successBlock:(void (^)(_WKFileUploadItem *))successBlock failureBlock:(void (^)(void))failureBlock
+{
+    if (NSString *uti = firstUTIThatConformsTo(result.itemProvider.registeredTypeIdentifiers, UTTypeMovie)) {
+        [result.itemProvider loadFileRepresentationForTypeIdentifier:uti completionHandler:^(NSURL *url, NSError *error) {
+            if (error) {
+                failureBlock();
+                return;
+            }
+
+            if (![url isFileURL]) {
+                LOG_ERROR("WKFileUploadPanel: Expected media URL to be a file path, it was not");
+                ASSERT_NOT_REACHED();
+                failureBlock();
+                return;
+            }
+
+            auto [maybeMovedURL, temporaryURL] = [self _copyToNewTemporaryDirectory:url];
+            self->_temporaryUploadedFileURLs.append(WTFMove(temporaryURL));
+
+            successBlock(adoptNS([[_WKVideoFileUploadItem alloc] initWithFileURL:maybeMovedURL.get()]).get());
+        }];
+
+        return;
+    }
+
+    NSString *uti = firstUTIThatConformsTo(result.itemProvider.registeredTypeIdentifiers, UTTypeImage);
+
+    if (!uti) {
+        LOG_ERROR("WKFileUploadPanel: Unexpected media type. Expected image or video, got: %@", mediaType);
+        ASSERT_NOT_REACHED();
+        failureBlock();
+        return;
+    }
+
+    [result.itemProvider loadFileRepresentationForTypeIdentifier:uti completionHandler:^(NSURL *url, NSError *error) {
+        if (error) {
+            failureBlock();
+            return;
+        }
+
+        if (![url isFileURL]) {
+            LOG_ERROR("WKFileUploadPanel: Expected media URL to be a file path, it was not");
+            ASSERT_NOT_REACHED();
+            failureBlock();
+            return;
+        }
+
+        auto [maybeMovedURL, temporaryURL] = [self _copyToNewTemporaryDirectory:url];
+        self->_temporaryUploadedFileURLs.append(WTFMove(temporaryURL));
```

Comment:
Does anything guarantee that `self` is still alive when this block is called?

Reply:
Doesn't the block implicitly retain `self`?